src/hooks/useResponsiveValue.ts

```diff
@@ -0,0 +1,60 @@
+import {useMedia} from './useMedia'
+
+// TODO: Write tests for this hook
+
+export type ResponsiveValue<TRegular, TNarrow = TRegular, TWide = TRegular> = {
+  narrow?: TNarrow // Applies when viewport is < 768px
+  regular?: TRegular // Applies when viewports is >= 768px
+  wide?: TWide // Applies when viewports is >= 1400px
+}
+
+/**
+ * Flattens all possible value types into single union type
+ * For example, if `T` is `'none' | 'line' | Responsive<'none' | 'line' | 'filled'>`,
+ * `FlattenResponsiveValue<T>` will be `'none' | 'line' | 'filled'`
+ */
+export type FlattenResponsiveValue<T> =
+  | (T extends ResponsiveValue<infer TRegular, infer TNarrow, infer TWide> ? TRegular | TNarrow | TWide : never)
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  | Exclude<T, ResponsiveValue<any>>
+
+/**
+ * Checks if the value is a responsive value.
+ * In other words, is it an object with viewport range keys?
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function isResponsiveValue(value: any): value is ResponsiveValue<any> {
+  return typeof value === 'object' && Object.keys(value).some(key => ['narrow', 'regular', 'wide'].includes(key))
+}
+
+/**
+ * Resolves responsive values based on the current viewport width.
+ * For example, if the current viewport width is less than 768px, the value of `{regular: 'foo', narrow: 'bar'}` this hook will return `'bar'`.
+ */
+export function useResponsiveValue<T, F>(value: T, fallback: F): FlattenResponsiveValue<T> | F {
+  // Check viewport size
+  // TODO: Get these breakpoint values from primer/primitives
+  const isNarrowViewport = useMedia('(max-width: 767px)') // < 768px
+  const isRegularViewport = useMedia('(min-width: 768px)') // >= 768px
```

Comment:
For high-density displays, we can't leave a 1px gap between `< 767` and `>= 768px`. That's why we use `0.02px` in our calculations.

See https://github.com/primer/css/blob/main/src/layout/page-layout.scss#L185